# Exercise 7: CNN for Cats and Dogs Data set

# Objective : Learning the application of CNN for image classification using raw data

# Problem : Construct a CNN for cats and dogs data set and compute the accuracy score

# Step 1: Code for uploading a zipfile and save it in a directory

In [ ]:
from google.colab import files
from io import BytesIO
uploaded=files.upload()

Saving cats_dogs.zip to cats_dogs.zip


In [ ]:
import io
import zipfile
data=zipfile.ZipFile(io.BytesIO(uploaded['cats_dogs.zip']))
data.extractall()
data.printdir()
data.close()

# Step 2: Create working folders in colab

In [ ]:
import os
base_dir='cats_dogs'
train_dir=os.path.join(base_dir,'train_dir')
test_dir=os.path.join(base_dir,'test_dir')
train_cats_dir=os.path.join(train_dir,'train_cats')
train_dogs_dir=os.path.join(train_dir,'train_dogs')
test_cats_dir=os.path.join(test_dir,'test_cats')
test_dogs_dir=os.path.join(test_dir,'test_dogs')

In [ ]:
#os.listdir(test_cats_dir) # to see the contents of a folder

# Step 3:  Data Generation

### Step 1 : Rescaling the Data

In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
train_datagen=ImageDataGenerator(rescale=1/255)
test_datagen=ImageDataGenerator(rescale=1/255)

### Step 2 : Training and Test Data Generation

In [ ]:
train_data=train_datagen.flow_from_directory(train_dir,target_size=(150,150),class_mode='binary')

Found 30 images belonging to 2 classes.


In [ ]:
test_data=test_datagen.flow_from_directory(test_dir,target_size=(150,150),class_mode='binary')

Found 20 images belonging to 2 classes.


### Step 3 : Storing the data and target values

In [ ]:
X,y=train_data.next()

**Getting some insight about X and y**

In [ ]:
print(X.shape)
print(y.shape)

In [ ]:
train_data.class_indices  # to know the labels of the classes

In [ ]:
print(y)

In [ ]:
X[1,].shape # giving the shape of the second image

(150, 150, 3)

In [ ]:
print(X[1,]) # displaying the content of image

# Step 4: Design of CNN for cats and dogs data

In [ ]:
from keras.models import Sequential
from keras.layers import Conv2D,MaxPool2D, Flatten, Dense

In [ ]:
model= Sequential()
model.add(Conv2D(32,kernel_size=(3,3),padding='same',activation='relu',input_shape=(150,150,3)))
model.add(MaxPool2D(pool_size=2))
model.add(Conv2D(64,kernel_size=(3,3),padding='same',activation='relu'))
model.add(MaxPool2D(pool_size=2))
model.add(Conv2D(128,kernel_size=(3,3),padding='same',activation='relu'))
model.add(MaxPool2D(pool_size=2))
model.add(Flatten())
model.add(Dense(64,activation='relu')) # dense hidden layer before the output layer
model.add(Dense(1,activation='sigmoid')) 
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_1 (Conv2D)           (None, 150, 150, 32)      896       
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 75, 75, 32)       0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 75, 75, 64)        18496     
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 37, 37, 64)       0         
 2D)                                                             
                                                                 
 conv2d_3 (Conv2D)           (None, 37, 37, 128)       73856     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 18, 18, 128)     

# Step 5: Compilation of the model

In [ ]:
model.compile(optimizer='rmsprop',loss='binary_crossentropy',metrics='accuracy')

# Step 6: Fitting of the model

In [ ]:
model.fit(X,y,epochs=50)

In [ ]:
import numpy as np
y_pr=np.round(model.predict(X))
y_pr=y_pr.reshape((30,))
print("Predicted classes",y_pr)
print("Target  classes",y)

1/1 [==============================] - 1s 823ms/step
Predicted classes [0. 0. 1. 1. 0. 1. 0. 1. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1. 1. 1. 0. 0. 1.
 0. 0. 0. 1. 0. 1.]
Target  classes [0. 0. 1. 1. 0. 1. 0. 1. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1. 1. 1. 0. 0. 1.
 0. 0. 0. 1. 0. 1.]


# Step 7: Prediction for test data

In [ ]:
X_test,y_test=test_data.next()
y_pr=np.round(model.predict(X_test))
y_pr=y_pr.reshape((20,))
print("Predicted classes",y_pr)
print("Target  classes",y_test)

1/1 [==============================] - 0s 274ms/step
Predicted classes [0. 1. 0. 1. 0. 0. 0. 1. 0. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1. 1.]
Target  classes [0. 1. 1. 1. 1. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 1. 0.]


In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_pr,y_test)

array([[5, 6],
       [5, 4]])

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_pr,y_test)

0.45

# Conclusion: 

## On predicting the 20 images, we obtain the confusion matrix of 
### [5, 6],
### [5, 4]

### The accuracy score is 0.45 as we trained the model with the less epochs. If we trained the model with more than 1000 epochs, we would have get more accuracy.